lstore/query.py

```diff
@@ -59,33 +62,62 @@ def insert(self, *columns):
     """
     def select(self, key, column, query_columns):
-        # 1. Go to specifed column and find the value that matches key
-        # 2. Check the query_columns list for the indexes set to 1
-        # 3. Check the schema encoding to see if we need to get the value from the base page or from the tail page
-        # 4. Aggregate all the values by the specifed query_columns list and return
-        # In milestone 1, this only returns a list with a single record
-        # EX: select(9898798, 0, [0, 1, 0, 1]) -> return [[89, 76]] 89 and 76 are from column 1 and 3 respectively 
-        # and 9898798 is the unique identifier (key) for that record
-        pass
+        # TODO record_does_exist(key) -> return rid if true and return false
+        # TODO  # read_record(key) -> returns a record if successful, else returns false
+                # Check the schema encoding to see if we need to get the value from the base page or from the tail page
+                # When you return a record, we need the MRU, not the value in the base page if an update exists for a column
+        selected_record = self.table.read_record(key)
+
+        if selected_record == False:
+            return False
+
+        return_list = []
+        filtered_record_list = []
+        for i in range(len(query_columns)):
+            if query_columns[i] == 1:
+                filtered_record_list.append(selected_record.user_data[i])
+            elif query_columns[i] == 0:
+                continue
+            else:
+                raise ValueError(f'ERROR: query_columns list must contain 0 or 1 values. {e}')
+        
+        return_list.append(filtered_record_list)
+        return return_list
 
     """
     # Update a record with specified key and columns
     # Returns True if update is succesful
     # Returns False if no records exist with given key or if the target record cannot be accessed due to 2PL locking
     """
     def update(self, key, *columns):
-        # Check if column has ever been updated and create a snapshot if it has not
+        # TODO : Discuss snapshots for future milestones
         # all updates will go to the tail page
-        # 1. Check that key exists in column 0 else return false
-        # 2. Create a new tail record, with a new tid
-        # 3. We will be given a list [None, None, NewValue, ... None, NewValue]
-        # 4. Update schema encoding at all columns with non None values to equal 1
-        # 5. Find the MRU and make a copy
-        # 6. Update base page indirection pointer to point to the new MRU
-        # 7. Update tail page indirection pointer of MRU to the SMRU
-        pass
+        if self.table.record_does_exist(key) == False:
+            return False
+        
+        current_record = self.table.read_record(key)
+        updated_schema_encoding = copy.deepcopy(current_record.schema_encoding)
+        updated_user_data = copy.deepcopy(current_record.user_data)
+        
+        for i in range(len(columns)):
```

Comment:
We can potentially offload this logic into the update function in Table if you like

Reply:
I think it's okay to leave it at this level because if not we have to pass in the columns and our idea was to keep read and write record more generic so we can use it in multiple places. We can move it in future milestones if it works better to have this internally.